File: Java/239.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    // not mine approach
       public int[] maxSlidingWindow(int[] nums, int k) {
        if(nums.length*k == 0) return new int[0];
        if(k ==1) return nums;
        
        int[] res = new int[nums.length - k + 1];
        int max = nums[0];
        for(int i = 0; i < k; i++){
            if(nums[i]>max) max = nums[i];
        }
        res[0] = max;
        int index = 0;
        for(int i = k; i<nums.length; i++){
            index = i-k+1;
            if(nums[i] > res[index - 1]){
                res[index] = nums[i];
            }else if(nums[index-1]!=res[index-1]){
                res[index] = res[index - 1];
            }else{
                max = nums[i];
                for(int j = index; j <i ; j++){
                    if(nums[j]>max) max = nums[j];
                }
                res[index] = max;
            }
        }
        return res;
    }
}
__________________________________________________________________________________________________
sample 37852 kb submission
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        if (n == 0) {
            return nums;
        }
        int[] result = new int[n - k + 1];
        LinkedList<Integer> dq = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (!dq.isEmpty() && dq.peek() < i - k + 1) {
                dq.poll();
            }
            while (!dq.isEmpty() && nums[i] >= nums[dq.peekLast()]) {
                dq.pollLast();
            }
            dq.offer(i);
            if (i - k + 1 >= 0) {
                result[i - k + 1] = nums[dq.peek()];
            }
        }
        return result;
    }
}
__________________________________________________________________________________________________

File: Python3/239.py
__________________________________________________________________________________________________
sample 152 ms submission
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if k == 0:
            return []
        if k == 1:
            return nums
        L = len(nums)
        res = [max(nums[:k])]
        for i in range(k, L):
            if nums[i] >= res[-1]:
                res.append(nums[i])
            else:
                if nums[i - k] < res[-1]:
                    res.append(res[-1])
                elif nums[i - k] == res[-1]:
                    res.append(max(nums[i-k+1:i+1]))
        return res
__________________________________________________________________________________________________
sample 19308 kb submission
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if k == 0 or len(nums) == 0:
            return []
        if k == 1 :
            return nums
        n = len(nums)
        left = [float('-inf')] * n
        right = [float('-inf')] * n
        left[0], right[-1] = nums[0], nums[-1]
        for i in range(1, n):
            #from left to right
            if i % k == 0:
                left[i] = nums[i]
            else:
                left[i] = max(left[i - 1], nums[i])
            #form right to left
            j = n - 1 - i
            if (j + 1) % k == 0:
                right[j] = nums[j]
            else:
                right[j] = max(right[j + 1], nums[j])
        res = []
        for i in range(n - k + 1):
            res.append(max(right[i], left[i + k - 1]))
        return res
__________________________________________________________________________________________________
